SLE-884: Open in IDE from SonarCloud

thahnen opened the pull request:
## Summary

Some changes had to be made to the assistance for connection creation in order to support SonarCloud alongside SonarQube for the **Open in IDE** feature and automatic Connected Mode setup in this case.

Additionally, for the automatic Connected Mode part the user experience was enhanced by waiting for the connection synchronization (and therefore analysis readiness) to be finished. These were some slightly bigger changes that will only trigger when not yet connected and triggering **Open in IDE**. When already bound, the logic is the same and not changed from the implementation done for SonarQube.

## Testing

There are no integration tests for SonarCloud as it is not in production. It was tested manually. Issues as described [HERE](https://sonarsource.atlassian.net/browse/SLE-797) were not fixed with this PR; they will be tackled either next week or when the Sonar Solution Squad is available.

thahnen committed:
SLE-884: Open in IDE from SonarCloud

Some changes had to be done to the assistance of the connection creation in order to also work and warn properly for connections to SonarCloud.

nquinquenel commented on org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/job/OpenIssueInEclipseJob.java:
```diff
@@ -81,7 +90,43 @@ public OpenIssueInEclipseJob(OpenIssueContext context) {
+        new AwaitProjectConnectionReadyJob(project, dialog, 0)
+          .schedule(AwaitProjectConnectionReadyJob.SCHEDULE_TIMER_MS);
+        dialog.open();
+        if (dialog.cancelledByJob() || dialog.cancelledByUser()) {
```
Can you maybe log if the job got canceled by the user or not? The next message is ambiguous, whether it was a timeout or canceled.

thahnen replied:
Yes, did a debug log

nquinquenel commented on org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/jobs/AnalysisReadyStatusCache.java:
```diff
@@ -0,0 +1,52 @@
+ *  must be thread safe!
+ */
+public class AnalysisReadyStatusCache {
+  private static final ConcurrentMap<String, Boolean> analysisReadyByConfigurationScopeId = new ConcurrentHashMap<>();
```
```suggestion
  private static final Map<String, Boolean> analysisReadyByConfigurationScopeId = new ConcurrentHashMap<>();
```
It might be down to preference, but we usually mention `Map` and not `ConcurrentMap`

thahnen replied:
Done

nquinquenel commented on org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/jobs/AnalyzeProjectJob.java:
```diff
@@ -200,7 +189,8 @@ private boolean checkIfReady() {
     for (var configurationScopeId : configurationScopeIds) {
-      if (Boolean.TRUE.equals(analysisReadyByConfigurationScopeId.get(configurationScopeId))) {
+      var readyness = AnalysisReadyStatusCache.getAnalysisReadiness(configurationScopeId);
+      if (Boolean.TRUE.equals(readyness)) {
```
```suggestion
      if (readyness) {
```

thahnen replied:
Thanks for the comment, due to this I found a bug here :smile: We were only waiting for one project to be ready and not for all of them...

thahnen committed:
SLE-893: Handle setting up Connected Mode automatically

When running Open in IDE (no matter of SonarQube or SonarCloud) and no connection / binding is established, we have to do it first. When the feature is then invoked afterwards, we have to await the connection to become ready (including analysis and branch matching) before we can proceed.
We want to provide the user with information and a more clear UX.

nquinquenel reviewed (approved):
LGTM!
